### PR TITLE
Fix hardcoded rocm warp size

### PR DESCRIPTION
**Pull request by petrex**
Update warp size constants(32) with __AMDGCN_WAVEFRONT_SIZE for ROCm

fix the reduction in `caffe2/sgd/adagrad_fused_op_gpu.cuh` ( not sure if ppl still use this kernel)
expect perf uplift for other cases

```
constexpr int kWarpSize = 32;

template <typename T>
inline __device__ T shfl_xor(const T val, int laneMask, int width = kWarpSize) {
#if !defined(USE_ROCM)
  return __shfl_xor_sync(0xffffffff, val, laneMask, width);
#else
  return __shfl_xor(val, laneMask, width);
#endif
}

/// Sums a register value across all warp threads
template <typename T, int ReduceWidth = kWarpSize>
inline __device__ T warpReduceAllSum(T val) {
#pragma unroll
  for (int mask = ReduceWidth / 2; mask > 0; mask >>= 1) {
    val += shfl_xor(val, mask);
  }
  return val;
}
```

Note that hardcoded warp size in `default_warp_iterator_from_smem.h` is not changed(not applicable to rocm)

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @XilunWu @tianyu-l @yf225 @chauhang

### CODE REVIEW

**Review by pruthvistony — APPROVED** [review]
I believe we can also use `warpSize` defined through hip_runtime.h - https://github.com/pytorch/pytorch/blob/main/c10/macros/Macros.h#L314
LGTM.

**Review by yifuwang — APPROVED** [review]
LG on torch/csrc/distributed/c10d/intra_node_comm.cu

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 1]
```diff
@@ -125,7 +125,11 @@ inline __host__ __device__ uint32_t getAlignmentRoundUp(const void* p) {
   return diff == 0 ? 0 : uint32_t(Align) - diff;
 }
 
+#if defined(USE_ROCM)
+constexpr int32_t kWarpSize = __AMDGCN_WAVEFRONT_SIZE;
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/int4mm.cu:129]: `warpSize` is a constexpr for HIP.  Prefer over internal `__AMDGCN_WAVEFRONT_SIZE`.

File: caffe2/utils/GpuDefs.cuh [hunk 2]
```diff
@@ -8,7 +8,7 @@ namespace caffe2 {
 // Static definition of GPU warp size for unrolling and code generation
 
 #if defined(USE_ROCM)
-constexpr int kWarpSize = warpSize;   // = 64 (Defined in hip_runtime.h)
+constexpr int kWarpSize = __AMDGCN_WAVEFRONT_SIZE;   // = 64 (Defined in hip_runtime.h)
```
Comment (jeffdaily) [inline caffe2/utils/GpuDefs.cuh:11]:
```suggestion
constexpr int kWarpSize = warpSize;   // = 64 (Defined in hip_runtime.h)
```
Please revert this change.

**Review by jeffdaily — CHANGES_REQUESTED** [review]
`warpSize` is a constexpr for HIP.  Prefer over internal `__AMDGCN_WAVEFRONT_SIZE`.  Be careful where `warpSize` is used.  Should only be used in device code.  If used in host code, the hip-clang compiler will default it to be 64.  Host code cannot use `warpSize` or any other constexpr to represent warpSize because it is dynamic and should be queried from device properties.

File: aten/src/ATen/native/cuda/int4mm.cu (hunk 1 above)
Comment (petrex) [reply aten/src/ATen/native/cuda/int4mm.cu:129]: thanks. Can you share how does that work?
quick grep but couldn't find that in ROCm/HIP...

Comment (jeffdaily) [reply aten/src/ATen/native/cuda/int4mm.cu:129]: https://github.com/ROCm/clr/blob/7ad141646420fb275d8d7358521f7806a92cda9a/hipamd/include/hip/amd_detail/amd_warp_functions.h#L76

__AMDGCN_WAVEFRONT_SIZE is defined by hip-clang compiler.  For device passes it will be set correctly based on gfx target.  For host passes, it defaults to 64.

Comment (jeffdaily) [reply aten/src/ATen/native/cuda/int4mm.cu:129]: https://clang.llvm.org/docs/HIPSupport.html#predefined-macros

Comment (petrex) [reply aten/src/ATen/native/cuda/int4mm.cu:129]: Thanks, so it's from clr. `static constexpr int warpSize = __AMDGCN_WAVEFRONT_SIZE;`

Let me switch to `warpSize` if thats preferred and fix the comments as well

**Review by eqy — APPROVED** [review]

**Comment by pruthvistony** [comment]
@petrex ,
Can you resolve the conflicts.

File: aten/src/ATen/native/cuda/int4mm.cu [hunk 3]
```diff
@@ -125,7 +125,11 @@ inline __host__ __device__ uint32_t getAlignmentRoundUp(const void* p) {
   return diff == 0 ? 0 : uint32_t(Align) - diff;
 }
 
+#if defined(USE_ROCM)
+constexpr int32_t kWarpSize = warpSize;
+#else
 constexpr int32_t kWarpSize = 32;
+#endif
```
Comment (jithunnair-amd) [inline aten/src/ATen/native/cuda/int4mm.cu:132]: `warpSize` is defined for CUDA as 32, so `if defined` is not required?

File: aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_backward.h [hunk 4]
```diff
@@ -242,8 +242,11 @@ struct AttentionBackwardKernel {
   static constexpr bool kKeysQueriesAlignedToBlockSize =
       kKeysQueriesAlignedToBlockSize_;
 
+#if defined(USE_ROCM)
+  static constexpr int64_t kWarpSize = warpSize;
+#else
   static constexpr int64_t kWarpSize = 32;
```
Comment (jithunnair-amd) [inline aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_backward.h:248]: warpSize is defined for CUDA as 32, so if defined is not required?

File: torch/csrc/distributed/c10d/intra_node_comm.cu [hunk 5]
```diff
@@ -10,7 +10,11 @@ namespace intra_node_comm {
 static constexpr size_t kBytesPerThread = 16;
 static constexpr size_t kMaxAllReduceBlocks = 24;
 static constexpr size_t kThreadsPerBlock = 1024;
+#if defined(USE_ROCM)
+static constexpr size_t kWarpSize = warpSize;
+#else
 static constexpr size_t kWarpSize = 32;
+#endif
```
Comment (jithunnair-amd) [inline torch/csrc/distributed/c10d/intra_node_comm.cu:17]: `warpSize` is defined for CUDA as 32, so `if defined` is not required?

File: aten/src/ATen/native/cuda/int4mm.cu (hunk 3 above)
Comment (petrex) [reply aten/src/ATen/native/cuda/int4mm.cu:132]: not sure I am following you? `warpSize` is defined in hip language runtime. We would need `else` for CUDA build

**Comment by jerrymannil** [comment]
> `warpSize` is a constexpr for HIP. Prefer over internal `__AMDGCN_WAVEFRONT_SIZE`. Be careful where `warpSize` is used. Should only be used in device code. If used in host code, the hip-clang compiler will default it to be 64. Host code cannot use `warpSize` or any other constexpr to represent warpSize because it is dynamic and should be queried from device properties.

@petrex 
In `int4mm.cu`, we use `warpSize` in host code to specify the kernel launch block dim.
See `launch_tinygemm_kernel` and `_convert_weight_to_int4pack_cuda`.
So this change will break things for int4 kernels

File: aten/src/ATen/native/cuda/int4mm.cu (hunk 3 above)
Comment (jerrymannil) [reply aten/src/ATen/native/cuda/int4mm.cu:132]: @petrex  See below comment.
We use it in host code

**Comment by pruthvistony** [comment]
> > `warpSize` is a constexpr for HIP. Prefer over internal `__AMDGCN_WAVEFRONT_SIZE`. Be careful where `warpSize` is used. Should only be used in device code. If used in host code, the hip-clang compiler will default it to be 64. Host code cannot use `warpSize` or any other constexpr to represent warpSize because it is dynamic and should be queried from device properties.
> 
> @petrex In `int4mm.cu`, we use `warpSize` in host code to specify the kernel launch block dim. See `launch_tinygemm_kernel` and `_convert_weight_to_int4pack_cuda`. I am adding further checks in by int4 PR. So you can leave it as-is for now

@petrex ,
For this code - https://github.com/pytorch/pytorch/blob/d8358a2d86de56945234744bf4721d7d548ba7ff/aten/src/ATen/native/cuda/int4mm.cu#L139
can we check on using the new parameter which is added to hipDeviceProperty to get the warpsize - https://github.com/pytorch/pytorch/pull/128449

**Comment by pruthvistony** [comment]
@petrex ,
As discussed changes are not required now, so closing the PR.
